networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java

```diff
@@ -171,4 +177,37 @@ private void createPersistentConnection(final PeerAddress peerAddress) {
                   TimeUnit.MILLISECONDS);
             });
   }
+
+  public long addPeerPredicate(final PeerPredicate predicate) {
+    return newPeerFilter.addPeerPredicate(predicate);
+  }
+
+  public boolean removePeerPredicate(final long id) {
+    return newPeerFilter.removePeerPredicate(id);
+  }
+
+  public static class NewPeerFilter {
+    private final AtomicLong predicateId = new AtomicLong();
+    private final Map<Long, PeerPredicate> peerPredicates = new ConcurrentHashMap<>();
+
+    long addPeerPredicate(final PeerPredicate predicate) {
+      final long id = predicateId.getAndIncrement();
+      peerPredicates.put(id, predicate);
+      return id;
+    }
+
+    boolean removePeerPredicate(final long predicateId) {
+      return peerPredicates.remove(predicateId) != null;
+    }
+
+    boolean isPeerValid(DiscoveryPeer peer) {
+      return peerPredicates.values().stream()
+          .map(predicate -> predicate.applyPeer(peer))
+          .reduce(true, (a, b) -> a && b);
```

Comment:
Nice.  But I think you can use:
```suggestion
      return peerPredicates.values().stream()
          .allMatch(predicate -> predicate.applyPeer(peer));
```

Reply:
Done.

Reply:
Done.